### ISSUE TITLE
Zig-zag artefact on spherical surface when manual texcoords computation is made

### ISSUE THREAD

**Issue by dsanhnt** [vispy/vispy#2330]
Hello everyone,

I'm working on the 3D visualization of planets in the solar system. As I am going to apply texture, I have manually computed the texture coordinates (texcoords) and I got [the zig-zag artefact as appears in the image](https://i.stack.imgur.com/yCOjd.png).

I believe that my computation might have something wrong. I have attached the texcoords computation below

```
# Compute Texture Coordinates  

def get_texcoords(vertices):
    texcoords = []
    for v in vertices:
        #thresholding
        for i in range(3):
            if np.abs(v[i]) > 1e-6:
                v[i] = v[i]
            elif np.abs(v[i]) < 1e-6:
                v[i] = 0.0

        # Compute position in uv-space
        radius = np.sqrt(v[0]**2 + v[1]**2 + v[2]**2)
        
        latitude = np.arcsin(v[2]/radius)
        longitude = np.arctan2(v[1],v[0])
            
        # Convert to texture coordinates
        u = round(0.5 + longitude/(2*np.pi),5)
        v = round(0.5 + latitude/np.pi,5)

        texcoords.append([u,v])
            
    return np.array(texcoords)
```
Is there any way to get those artefacts away, or does it have a smarter way to obtain texture coordinate in vispy.

Thank you for your help and suggestion.
Sanhanat

**Comment by kmuehlbauer** [comment]
@dsanhnt Maybe you can adapt from @rougier here: https://github.com/glumpy/glumpy/blob/master/examples/earth.py

**Comment by djhoese** [comment]
Did you post this on stackoverflow or on gitter too? I know I've seen someone talking about this issue already.

**Comment by dsanhnt** [comment]
> Did you post this on stackoverflow or on gitter too? I know I've seen someone talking about this issue already.

Yes, I posted it on StackOverflow last month, but there’s no reply.

**Comment by djhoese** [comment]
Sorry about that. For the record here is the question: https://stackoverflow.com/questions/71860484/how-can-i-get-rid-of-zig-zag-artifact-on-sphere-surface-in-vispy

I thought I had replied to that, but apparently not. My first question is why is the `round` needed on the texture coordinates? How are you using/applying the coordinates? Can you show us more of your code or produce a minimal example?

**Comment by dsanhnt** [comment]
First of all, I used `round()` to make some approximation, which I believed that it is precise enough for such a case like this. For your second question, my texcoords are computed based on existing vertices generated by the VisPy `get_vertices()` function and I apply them to geometry with the `TextureFilter()`. [The minimal example](https://github.com/vispy/vispy/files/8675363/demo_sphere_with_texture.zip) can be found as an attached zip file.

**Comment by rougier** [comment]
It looks like you might be off by one index. You should try the same with a grid texture (possibily with lables) to debug what is wrong.

**Comment by dsanhnt** [comment]
Thank you @rougier for your hint. one index is missing as you suspected. 

assuming n_columns = 20, 19 sectors will be normally rendered. There is only one sector at the border of an image that there has no point represented, due to my computation that uses corners of the triangle instead of the midpoint of the rectangle. Hence, vispy automatically fills the missing sector with the whole image.

I have changed the method of rendering from `latitude` to `ico`, then the artefact is changed from zig-zag to one appears on the image

![image](https://user-images.githubusercontent.com/37921133/169244945-a21a3c35-7bfa-4a55-af46-48f6673e9561.png)

**Comment by rougier** [comment]
Can you use a texture like: https://cs.wellesley.edu/~cs307/threejs/r67/examples/textures/UV_Grid_Sm.jpg. This generally helps to spot the error.

**Comment by dsanhnt** [comment]
@rougier it looks like the wrapping does not go completely around.
![image](https://user-images.githubusercontent.com/37921133/169466812-107b64b3-d384-4295-84ee-e844e18bb25c.png)

**Comment by djhoese** [comment]
It looks to me like the first portion of your texture coordinates are wrong, right? I'm not sure I understood your comment about why vispy would repeat the entire texture between these two portions, but we can see it clearly here. The left part of the texture is either not being used or is being corrupted/overwritten.

This probably isn't anything to worry about or that has to do with this problem, but I did notice in your original code you have:

```python
        #thresholding
        for i in range(3):
            if np.abs(v[i]) > 1e-6:
                v[i] = v[i]
            elif np.abs(v[i]) < 1e-6:
                v[i] = 0.0
```

Isn't the first `if` unnecessary? Technically this if statement doesn't handle the case where abs(v) is exactly 1e-6 either.

**Comment by rougier** [comment]
This could help: https://localcoder.org/seam-issue-when-mapping-a-texture-to-a-sphere-in-opengl
